platform-sdk/docs/proposals/consensus-layer/Consensus-Layer.md

```diff
@@ -0,0 +1,438 @@
+It also makes sure they are durably persisted before emission so as to add resilience to the node in case of certain
+catastrophic failure scenarios.
+
+#### Validation
+
+One of the core responsibilities of Event Intake is to validate the events it has received. While this document does
+not specify the validation pipeline, it will define some of the primary steps involved in validation, so as to motivate
+the purpose of this module. That is, the following description is non-normative, but important for understanding the
+context within which this module operates.
+
+Event Intake receives events from gossip, or from the Event Creator module (i.e. a "self-event"). An "event" is actually
+a wrapper around the base protobuf event type (`EventCore`) that includes the core event data along with metadata that
+is not represented in protobuf or transmitted through gossip. One field populated in the metadata will be the byte[]
+that represents the protobuf-serialized form of the `EventCore`. Since gossip read this from the wire, there is no work
+to produce this `byte[]`, and since the `byte[]` will be needed for Gossip, we need to keep it anyway. But we will also
+hash this `byte[]` to produce the event hash, and store this hash in the metadata.
+
+After hashing, Event Intake will deduplicate events. It has a hash->event map allowing it to cheaply verify whether the
+event is a duplicate. If it is, the duplicate is discarded. Otherwise, the event is checked for "syntactic" correctness.
+For example, are all required fields populated, etc. While the Gossip system has already checked to ensure the payload
+of the event (its transactions) are limited in size and count, Event Intake will also check this as a safety measure in
+the event of bugs (if this check fails here, a noisy log statement should be produced, since this should never happen).
+
+If an event is valid, then we finally check the signature. Since validation and deduplication and hashing are
+significantly less expensive than signature verification, we wait on signature verification until the other steps are
+completed. The operating principle is that we want to fail fast and limit work for further stages in the pipeline.
+
+If an event has a very old birth-round that is expired, it is dropped. If a node sends a large number of expired events,
+it may end up being disciplined (the exact rules around this will be defined in subsequent design docs for the
+Event Intake module).
+
+#### Self Events
+
+Events are not only given to the Event Intake system through gossip. Internal events are also fed to Event Intake.
+These internal events **may** bypass some steps in the pipeline. For example, internal self-events (those events created
+by the node itself) do not need validation. Likewise, when replaying events from the pre-consensus event buffer, those
+checks are not needed (since they have already been proved valid and are in topological order).
```

Comment:
The correct term is "self event", not "internal event". This term shows up all over the place in variable names and documentation. Such a thing could be renamed, but if so it should probably be an intentional choice.